### PR TITLE
Don't run net4x tests with Mono on non-Windows

### LINKED ISSUE #81395: Running tests tries to run .NET Framework tests with mono on non-Windows
When I run tests on my Mac now, it seems to be trying to run .NET Framework tests with mono. For example, running this command on my Mac

```shell
./dotnet.sh test src/libraries/System.Security.Cryptography.Pkcs/tests
```

It will first run all of the .NET (Core) tests and everything passes. Great. Then it proceeds to run the .NET Framework test, and gives this error:

```
A total of 1 test files matched the specified pattern.
System.IO.FileNotFoundException: Could not find 'mono' host. Make sure that 'mono' is installed on the machine and is available in PATH environment variable.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Helpers.DotnetHostHelper.GetMonoPath()
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DefaultTestHostManager.GetTestHostProcessStartInfo(IEnumerable`1 sources, IDictionary`2 environmentVariables, TestRunnerConnectionInfo connectionInfo)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyExecutionManager.StartTestRun(TestRunCriteria testRunCriteria, IInternalTestRunEventsHandler eventHandler)
```

Even if I install mono, the tests will absolutely crash and burn since most of the project I work on that target .NET Framework assume they are running on Windows. Running S.S.C.Pkcs on macOS with Mono with the .NET Framework target has hundreds of failures.

Was there a change that introduced this recently, and, was it intentional? Or did something change in my environment that's making it try to do this?

### PR DESCRIPTION
**Pull request by akoeplinger**
We'd try to use Mono to run them but most of the tests assume Windows when targetting net4x so they don't work anyway and you'd need to remember to pass the `-f net9.0` TFM filter. Instead just skip them on non-Windows platforms.

Fixes https://github.com/dotnet/runtime/issues/81395

### CODE REVIEW

**Review by ViktorHofer — COMMENTED** [review]
Few questions / observations:
1. This only filters .NET Framework builds out for multi-targeting projects, right? How would we achieve the same when i.e. doing a `dotnet test` on a traversal project / solution file for non multi-targeting projects? While we only have a few of those, I think we should keep parity between single and multi-targeting builds in regards to test adapter support.
2. Won't this break whenever the SDK or this repo adds to the `AdditionalProperties` metadata? Is there a more robust way to make this comparison?
3. This reminds me that the current experience is the one that customer hit as well. I would prefer if we could fix this not just for runtime but for all other repositories and our customers as well. If we decide that we only want to fix this for runtime, this would benefit from a comment pointing to a tracking issue on the vstest side.

**Comment by akoeplinger** [comment]
> This only filters .NET Framework builds out for multi-targeting projects, right? How would we achieve the same when i.e. doing a dotnet test on a traversal project / solution file for non multi-targeting projects? While we only have a few of those, I think we should keep parity between single and multi-targeting builds in regards to test adapter support.

I don't know how it works for traversal projects or solution files, does it not run the Test target on the individual projects?
Not sure what you mean with non multi-targeting projects, do we have ones where we only target a net4x TFM?

> Won't this break whenever the SDK or this repo adds to the AdditionalProperties metadata? Is there a more robust way to make this comparison?

Yes. I considered a few options but they're all quite ugly and I figured having the risk of this breaking (no big deal, we'd just start running netfx targets again) is worth the simpler code. The SDK shouldn't be an issue since GetProjectWithBestTargetFrameworks is an arcade target.

> This reminds me that the current experience is the one that customer hit as well. I would prefer if we could fix this not just for runtime but for all other repositories and our customers as well. If we decide that we only want to fix this for runtime, this would benefit from a comment pointing to a tracking issue on the vstest side.

For customers the existing behavior makes sense, or at least it'd be a breaking change. In dotnet/runtime we can make different assumptions.

**Comment by ViktorHofer** [comment]
I wasn't aware that this only affects projects that use the `Microsoft.DotNet.Buid.Tasks.TargetFramework` nuget package. I think it would be better to move this into here: https://github.com/dotnet/arcade/blob/130b08562936277405284145c1befc87bf1e83cb/src/Microsoft.DotNet.Build.Tasks.TargetFramework/src/buildMultiTargeting/Microsoft.DotNet.Build.Tasks.TargetFramework.targets#L40-L41

By doing that, any consumer of that nuget package will benefit from your improvements.

We already check the TFM values there to not append `-$(TargetOS)` for .NET Framework so that would be a good place to filter those projects out on non Windows. If you make that the default, please add an opt-out switch for the new behavior.

As discussed offline, we should still file an issue on the VSTest side and link to it for the `dotnet test` code path.

**Comment by ViktorHofer** [comment]
I wasn't aware that this only affects projects that use the `Microsoft.DotNet.Buid.Tasks.TargetFramework` nuget package. I think it would be better to move this into here: https://github.com/dotnet/arcade/blob/130b08562936277405284145c1befc87bf1e83cb/src/Microsoft.DotNet.Build.Tasks.TargetFramework/src/buildMultiTargeting/Microsoft.DotNet.Build.Tasks.TargetFramework.targets#L40-L41

By doing that, any consumer of that nuget package will benefit from your improvements. And the approach is more efficient as we don't need to pass the to be filtered project TFM tuples into the `ChooseBestTargetFrameworksTask ` task.

We already check the TFM values there to not append `-$(TargetOS)` for .NET Framework so that would be a good place to filter those projects out on non Windows. If you make that the default, please add an opt-out switch for the new behavior.

As discussed offline, we should still file an issue on the VSTest side and link to it for the `dotnet test` code path.

**Comment by akoeplinger** [comment]
@ViktorHofer updated with the new arcade-based filtering. Also filed a VSTest issue: https://github.com/microsoft/vstest/issues/4908

File: eng/testing/outerBuild.targets [hunk 1]
```diff
@@ -1,12 +1,18 @@
 <Project>
-  <Target Name="Test" DependsOnTargets="GetProjectWithBestTargetFrameworks">    
+  <Target Name="Test" DependsOnTargets="RemoveNetFrameworkProjectsOnNonWindows;GetProjectWithBestTargetFrameworks">
     <MSBuild Projects="@(InnerBuildProjectsWithBestTargetFramework)"
              Targets="Test">
     </MSBuild>
   </Target>
-  <Target Name="VSTest" DependsOnTargets="GetProjectWithBestTargetFrameworks">    
+  <Target Name="VSTest" DependsOnTargets="RemoveNetFrameworkProjectsOnNonWindows;GetProjectWithBestTargetFrameworks">
```
Comment (ViktorHofer) [inline eng/testing/outerBuild.targets:7]:
```suggestion
  <!-- Manually remove .NET Framework on non-Windows: https://github.com/microsoft/vstest/issues/4908 -->
  <Target Name="VSTest" DependsOnTargets="RemoveNetFrameworkProjectsOnNonWindows;GetProjectWithBestTargetFrameworks">
```

**Review by ViktorHofer — APPROVED** [review]
One nit, aside from that LGTM. Regarding the portable source build failure, tt might not be possible anymore to just update one dependency from a repo that offers an intermediate source build package.

**Comment by akoeplinger** [comment]
> Regarding the portable source build failure, tt might not be possible anymore to just update one dependency from a repo that offers an intermediate source build package.

Ok, I'll wait for the arcade update to flow into main.

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**